pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1385,4 +1386,89 @@ public void uploadToDeepStoreIfMissing(TableConfig tableConfig, List<SegmentZKMe
+    streamConfig.setOffsetCriteria(OffsetCriteria.SMALLEST_OFFSET_CRITERIA);
+    List<PartitionGroupMetadata> newPartitionGroupMetadataList =
+        getNewPartitionGroupMetadataList(streamConfig, currentPartitionGroupConsumptionStatusList);
+    streamConfig.setOffsetCriteria(originalOffsetCriteria);
+
+    int numPartitionGroups = newPartitionGroupMetadataList.size();
+    Map<Integer, SegmentZKMetadata> partitionToLatestSegment = getLatestSegmentZKMetadataMap(realtimeTableName);
+    for (PartitionGroupMetadata partitionGroupMetadata : newPartitionGroupMetadataList) {
+      int partitionId = partitionGroupMetadata.getPartitionGroupId();
+      if (partitionToLatestSegment.containsKey(partitionId)
+          && idealState.getInstanceStateMap(partitionToLatestSegment.get(partitionId).getSegmentName())
+              .containsValue(SegmentStateModel.CONSUMING)) {
+        // IN_PROGRESS segment already exists, NO-OP
+        LOGGER.info("Skipping partitionGroupId {}, IN_PROGRESS segment already exists", partitionId);
+      } else {
+        // Step 1: Create PROPERTYSTORE nodes for each partitionId (IN_PROGRESS)
+        String offset = partitionGroupMetadata.getStartOffset().toString();
+        long newSegmentCreationTimeMs = System.currentTimeMillis();
+
+        SegmentZKMetadata latestSegmentMeta = partitionToLatestSegment.get(partitionId);
+        int seqNumber = latestSegmentMeta == null ? STARTING_SEQUENCE_NUMBER
+            : new LLCSegmentName(latestSegmentMeta.getSegmentName()).getSequenceNumber() + 1;
+
+        LLCSegmentName newLLCSegment = new LLCSegmentName(tableName, partitionId, seqNumber, newSegmentCreationTimeMs);
+        SegmentZKMetadata newSegmentZKMetadata = new SegmentZKMetadata(newLLCSegment.getSegmentName());
+        newSegmentZKMetadata.setCreationTime(newLLCSegment.getCreationTimeMs());
+        newSegmentZKMetadata.setStartOffset(offset);
+        newSegmentZKMetadata.setNumReplicas(numReplicas);
+        newSegmentZKMetadata.setStatus(CommonConstants.Segment.Realtime.Status.IN_PROGRESS);
+        SegmentPartitionMetadata segmentPartitionMetadata =
+            getPartitionMetadataFromTableConfig(tableConfig, partitionId);
+        if (segmentPartitionMetadata != null) {
+          newSegmentZKMetadata.setPartitionMetadata(segmentPartitionMetadata);
+        }
+
+        FlushThresholdUpdater flushThresholdUpdater = _flushThresholdUpdateManager
+            .getFlushThresholdUpdater(streamConfig);
+        CommittingSegmentDescriptor committingSegmentDescriptor = new CommittingSegmentDescriptor(null, offset, 0);
+        flushThresholdUpdater.updateFlushThreshold(streamConfig,
+            newSegmentZKMetadata, committingSegmentDescriptor, null,
+            getMaxNumPartitionsPerInstance(instancePartitions, numPartitionGroups, numReplicas),
+            newPartitionGroupMetadataList);
+
+        _helixResourceManager.getPropertyStore().set(
```

Comment:
we have to ensure that this change plays nicely with the code in ensureAllPartitionsConsuming. That method also creates new zkMetadata and idealState entry, and these 2 should not interfere with each other.  You'll have to add some synchronization on table name between the 2 methods

Reply:
With the redesign, this is achieved. Although I had to serialize calls to `ensureAllPartitionsConsuming` since there is a possibility of API call running concurrently with the periodic task. I think we can improve upon this by having table level locks being acquired within the function. Have added that as a todo since it should not be much of an issue I feel.

Reply:
Makes sense. TableName level locks won't help much in that case. But synchronizing the method is still needed? To address the, albeit rare, possibility of the API being hit on the lead controller, while the periodic task is running.
Since its a background periodic task, the overhead of the syncronized function shouldn't be much I feel.. Any suggestions on this?

Reply:
synchronization is still needed for same controller yes.. but the other case is more concerning. for instance, on the 2 controllers (1 via periodic task thread other via resumeConsumption thread), we could enter setupNewPartitionGroup method at once, and end up with 2 zk metadata and 2 CONSUMING segments. Many more such cases could come up.
Wondering now, if adding a back door entry to this method from any controller is a good idea. This won't happen if only periodic task is allowed to execute that method (be it scheduled or manual). So 1) should this flag just be part of manual periodic task trigger options and then 2) this API internally just invokes the periodic task

Reply:
I agree. If this method is only called via the periodic task - RealtimeSegmentValidationManager - then we could avoid these potential race problems. The new flag needs to be wired to the periodic task though. Looked at the code for manually running the periodic task. `PinotControllerPeriodicTaskRestletResource.runPeriodicTask()` can be refactored to take task parameters...

Reply:
I was trying to go through helix's `updateIdealState` function, which gets called by `ensureAllPartitionsConsuming`, and looks like the idealState update has support for handling concurrent writes (CAS like implementation). So we may not need syncronization afterall? Even if two threads / processes try to concurrently update the state, it'll get serialized and hence will be safe.
 CC: @Jackie-Jiang for more inputs

Reply:
@saurabhd336 (cc @Jackie-Jiang) If it was just ideal state changes, what you're saying would be fine, but `ensureAllPartitionsConsuming` method also adds segmentZkMetadata and that part is not serialized!
Manually kicking off RealtimeSegmentValidation job takes care of both ideal states and segZkMetadata changes as it serializes running of the whole job so that only one validation job is running at any given time.

Reply:
@sajjad-moradi It should be okay because we always follow the same sequence of operation as the segment commit protocol:
```
Step 1: Update PROPERTYSTORE to change the old segment metadata status to DONE
Step 2: Update PROPERTYSTORE to create the new segment metadata with status IN_PROGRESS
Step 3: Update IDEALSTATES to include new segment in CONSUMING state, and change old segment to ONLINE state.
```
The operation itself is idempotent. Given any intermediate state, it will return the same result.

If the operation is not idempotent, then we also need to handle the synchronization across multiple controllers.

Reply:
Step 2 is not idempotent. Because of the timestamp in the new segment name, everytime step 2 gets executed, we'll get a new segment name which leads to new segZkMetadata:
```java
      long newSegmentCreationTimeMs = getCurrentTimeMs();
      LLCSegmentName newLLCSegment = new LLCSegmentName(rawTableName, committingSegmentPartitionGroupId,
          committingLLCSegment.getSequenceNumber() + 1, newSegmentCreationTimeMs);
      createNewSegmentZKMetadata(tableConfig, streamConfig, newLLCSegment, newSegmentCreationTimeMs,
          committingSegmentDescriptor, committingSegmentZKMetadata, instancePartitions, numPartitionGroups, numReplicas,
          newPartitionGroupMetadataList);

```

Reply:
Good point. So it can potentially lead to race condition if 2 runs happen at exactly the same time, and both create the new segment ZK metadata. One of the segment ZK metadata will be picked and continue, the other will be left orphan. 
@saurabhd336 Let's change `ensureAllPartitionsConsuming` to `synchronized`, and add a TODO describing this potential race condition. We should probably clean up the orphan segment ZK metadata if it is not in the ideal state. This can be handled in a separate PR.

Reply:
We don't need to do so if we go with manually kicking RealtimeSegmentValidation job as Neha suggested. The code in `PinotControllerPeriodicTaskRestletResource` can be modified a bit to address to achieve this.

Reply:
@sajjad-moradi Not sure I understand 'manually kicking RealtimeSegmentValidation job'. That still means `ensureAllPartitionsConsuming` can get called concurrently by the API thread and the scheduled thread?

Reply:
@saurabhd336 just think of the periodic task. There's an API to invoke periodic task manually (look for Periodic Task tab in swagger). Whether it runs by schedule, or by manual trigger, periodic task can only run on lead controller of a table, so that eliminates across controller clashes. As for same controller, if you look at BasePeriodicTask, you'll see it doesn't let more than one task to run at once.

So the suggestion is, rely on this mechanism as we already know this works and prevents clashes. One way to rely on this mechanism is by introducing a query param to the runPeriodicTask API with our extra config and pass it to the periodic task execution. Another way could be to keep your new API, but again use same mechanism happening in runPeriodicTask API to trigger the task.

does this help clarify? We can chat on the slack channel if need be.

Reply:
I did not have full context on the periodic task execution via API. But now I've updated the PR with
1) Allowing custom params to be passed to on demand task execution via helix messages.
2) Changed the API to use the periodic task execution instead of directly calling the method. 
Please have a look @npawar @Jackie-Jiang @sajjad-moradi